Stats: Hide search terms module on Traffic page

a8ck3n opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to:

https://github.com/Automattic/jetpack/issues/39919

## Proposed Changes

Hides the search terms module on the Traffic page by default and adds a new toggle to the visibility settings so people can keep it visible if they want.

### The Search terms module
<img width="748" alt="SCR-20250114-ncps" src="https://github.com/user-attachments/assets/47c833df-f811-495e-bfa8-f96bb9024c48" />

### New visibility toggle
<img width="739" alt="SCR-20250120-sjrx" src="https://github.com/user-attachments/assets/9e59bcb6-df8a-4199-b968-8ccc3254ef31" />

### Bug fixes
Also fixes two bugs where defaults were being ignored, one on the Traffic page and one in the `PageModuleToggler` component. The result was that any module without a saved state (ie: the toggle had never been used) would be visible on the page by default but the settings would show it as disabled. This was easily reproducible on new sites.

One other bug was fixed in the derived state of the `StatsNavigation` component. The logic there was updating all toggles based on the state of the VideoPress feature. So a site with VideoPress disabled would force all other toggles to enabled while a site with VideoPress enabled would force all other toggles to disabled.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

As it currently works, the module displays an incomplete data set and introduces confusion for users who might have insights via Google Analytics, potentially reducing trust with users.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Load the Traffic page and confirm the "Search terms" module is no longer visible by default.
- Test at different screen widths to verify no layout issues are introduced.
- Open the visibility settings and confirm the toggle is off by default.
- Confirm the module is shown/hidden to match the state of the toggle.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

a8ck3n committed:
Hide the search terms module on the Traffic page

a8ck3n committed:
Comments++

annacmc reviewed (approved):
Looks good and works as expected to me. Couldn't find any layout issues either. I like how you've kept the route active, I agree this will make testing much easier.

kangzj reviewed (commented):
I would prefer to still allowing the toggling of the module in settings; otherwise we'd probably expect quite a few complaints coming through.

a8ck3n commented:
> I would prefer to still allowing the toggling of the module in settings; otherwise we'd probably expect quite a few complaints coming through.

I'm looking into adding this to the toggle now. I've run into a few bugs with the current setup though. I'm going to take a quick look and try to address those first. The main issue I'm seeing is that for sites who have never clicked on the toggles, they do not match or respect the defaults. It differs based on site type/environment but one example is that a new Simple site without a plan should not show either the Authors or Videos modules by default. The settings popup shows both as being disabled while the page still shows them. I thought this was just an issue with the page but there are bugs on both sides (the page display logic and the toggle display logic).

Anyway, will try to address those as well as get the Search term module into the toggles.

a8ck3n committed:
Add Search module to visibility toggles

a8ck3n committed:
Optionally present Search terms module

a8ck3n committed:
Apply defaults to visibility settings

a8ck3n committed:
Memoize derived visibility state

a8ck3n committed:
Fix bug in derived state

Previous logic would be applied to all toggles which isn’t what we want.

a8ck3n committed:
Handle both null and undefined

Previous check was explicit but the value could also be undefined if the toggle had never been used. This would cause the default value to be ignored and always return false for the toggle state.

a8ck3n committed:
Disable VideoPress override

a8ck3n commented on client/blocks/stats-navigation/index.js:
```diff
@@ -95,13 +95,19 @@ class StatsNavigation extends Component {
+					};
+				}
 
-				return {
```
Previous logic here was being applied to every `toggleItem` which made it impossible to have an item disabled by default. New logic only updates state for the `videos` item. Currently disabled based on the `optionallyDisableVideoPress ` boolean value.

a8ck3n commented on client/blocks/stats-navigation/constants.ts:
```diff
@@ -32,6 +32,15 @@ export const AVAILABLE_PAGE_MODULES: Record< string, ModuleToggleItem[] > = {
+			},
+			icon: search,
+			defaultValue: false,
+			disabled: false,
```
The `disabled` value is currently ignored by the Settings toggle. This is probably fine as I don't think we need to force anything to be disabled for now.

kangzj replied:
Ya it was done intentionally here https://github.com/Automattic/wp-calypso/pull/97913 - basically we shouldn't disable it if we don't have full confidence I guess.

a8ck3n commented on client/blocks/stats-navigation/index.js:
```diff
@@ -95,13 +95,19 @@ class StatsNavigation extends Component {
+				// TODO: Review this logic.
+				// By default, all sites have the ability to upload a single video.
+				// Therefore, it probably doesn't make sense to disable the module by default.
+				const optionallyDisableVideoPress = false;
```
I'm disabling the VideoPress override pending feedback.

I'm of the opinion that we don't need this logic now as all sites have the ability to upload a single video. Therefore, I'm not sure if it makes sense to hide the module based on the `siteHasFeature()` check on line 307.

I'm thinking if VideoPress support is disabled, it's good enough to allow the user to toggle the Stats module off. Any thoughts on this one?

kangzj replied:
My intention was only to set a default value for the VideoPress module - if there is only one video, I don't know how much value the stats module would be tbh.

The changes were initially introduced here: https://github.com/Automattic/wp-calypso/pull/96894.

We should probably still default the toggle to false if site doesn't have subscription. Either way tho, I think it'll be better to exclude these changes from the PR except for fixing the logic here.

a8ck3n replied:
Okay, I'll give this another look. The only downside to keeping this behaviour is that we need to refactor it (or duplicate the checks) for the Traffic page. Otherwise the toggle and the Traffic page get out of sync.

a8ck3n replied:
I added a helper function that exposes the default module setup so it's easily shared between the two components making this less of a worry now.

a8ck3n commented on client/blocks/stats-navigation/index.js:
```diff
@@ -95,13 +95,19 @@ class StatsNavigation extends Component {
-					...toggleItem,
-					disabled: shouldDisableVideoToggle,
-					defaultValue: shouldDisableVideoToggle === false,
-				};
```
The previous logic was actually being applied to all toggles which isn't what we want. The end result was that it wasn't possible to have a toggle off by default. The new logic (currently disabled) is only applied to the Videos toggle.

a8ck3n commented on client/blocks/stats-navigation/page-module-toggler.tsx:
```diff
@@ -81,7 +81,7 @@ export default function PageModuleToggler( {
 									className="page-modules-settings-toggle-control"
 									checked={
-										pageModules[ toggleItem.key ] !== null
+										pageModules[ toggleItem.key ] != null
```
The API can return empty or partial results, for example: `{}` or `{ "authors": false }` so we need a "truthy" check to handle both `null` and `undefined` here. Otherwise these toggles will always show as disabled when a value hasn't been previously saved.

kangzj replied:
Nice catch! This one is on me.

~~we should probably change it to `pageModules[ toggleItem.key ] !== undefined` - I somehow thought if a value is not set, it would be `null`. What do you think?~~

Maybe just a truthy check `pageModules[ toggleItem.key ]` here?

a8ck3n replied:
So changing this check doesn't work as it evaluates the value instead of the presence of a value. When the value of `pageModules[ toggleItem.key ]` is false, the check will use the default value instead of the provided value.

I was a bit confused by this but CoPilot helped me understand the difference and recommend I use the "nullish coalescing operator" instead. Nice!

a8ck3n reviewed (commented):
A few comments on the changes.

kangzj commented on client/my-sites/stats/site.jsx:
```diff
@@ -176,6 +186,12 @@ function StatsBody( { siteId, chartTab = 'views', date, context, isInternal, ...
 	const momentSiteZone = useSelector( ( state ) => getMomentSiteZone( state, siteId ) );
 
+	// Determine module visibility based on user settings AND defaults.
+	const moduleVisibility = useMemo(
```
I like how the default values are handled here - elegant and high performance 👍

a8ck3n committed:
Update toggle defaults based on VideoPress state

a8ck3n committed:
Update Traffic page to check VideoPress state

a8ck3n committed:
Simplify truthy check for user preference

a8ck3n committed:
Refactor setup for default module listings

a8ck3n committed:
Update Traffic page to use default modules helper

a8ck3n committed:
Fix check for presence of value

a8ck3n committed:
Simplify evaluation of checked toggles

kangzj reviewed (approved):
Looks good to me! Nice and clean work 👍

kangzj committed:
Merge branch 'trunk' into update/hide-search-terms-module

kangzj commented on client/my-sites/stats/site.jsx:
```diff
@@ -6,12 +6,12 @@ import clsx from 'clsx';
 import { find } from 'lodash';
 import moment from 'moment';
-import { useState, useEffect, useCallback, useMemo } from 'react';
+import { useCallback, useState, useEffect, useCallback, useMemo } from 'react';
```
```suggestion
import { useState, useEffect, useCallback, useMemo } from 'react';
```

kangzj committed:
Update client/my-sites/stats/site.jsx
